src/engine/input-controllers/keyboard-input.ts

```diff
@@ -0,0 +1,74 @@
+import KeyboardMapping from './keyboard-mapping'
+import ArrayUtils from './../util/array-utils'
+import InputController from './input-controller'
+
+export default class KeyboardController extends InputController {
+
+  private mappings: Array<KeyboardMapping> = [];
+
+  private hasDownHandler: number = 0;
+  private hasUpHandler: number = 0;
+  private hasPressHandler: number = 0;
+
+  // These functions need to be wrapped to protect `this` context
```

Comment:
Yea it's weird.  this gets assigned the enclosing class the function is called in. Those function references end up getting called in another class.